src/Castle.Core/DynamicProxy/Generators/Emitters/AbstractTypeEmitter.cs

```diff
@@ -266,42 +264,72 @@ public FieldReference GetField(string name)
 			return value;
 		}
 
-		public Type GetGenericArgument(string genericArgumentName)
+		public Type GetClosedParameterType(Type parameter)
 		{
-			if (name2GenericType.TryGetValue(genericArgumentName, out var genericTypeParameterBuilder))
-				return genericTypeParameterBuilder;
+			if (parameter.IsGenericType)
+			{
+				// ECMA-335 section II.9.4: "The CLI does not support partial instantiation
+				// of generic types. And generic types shall not appear uninstantiated any-
+				// where in metadata signature blobs." (And parameters are defined there!)
+				Debug.Assert(parameter.IsGenericTypeDefinition == false);
 
-			return null;
-		}
+				var arguments = parameter.GetGenericArguments();
+				if (CloseGenericParametersIfAny(arguments))
+				{
+					return parameter.GetGenericTypeDefinition().MakeGenericType(arguments);
+				}
+			}
 
-		public Type[] GetGenericArgumentsFor(Type genericType)
-		{
-			var types = new List<Type>();
+			if (parameter.IsGenericParameter)
+			{
+				return GetGenericArgument(parameter.GenericParameterPosition);
+			}
 
-			foreach (var genType in genericType.GetGenericArguments())
+			if (parameter.IsArray)
 			{
-				if (genType.IsGenericParameter)
-				{
-					types.Add(name2GenericType[genType.Name]);
-				}
-				else
+				var elementType = GetClosedParameterType(parameter.GetElementType());
+				int rank = parameter.GetArrayRank();
+				return rank == 1
+					? elementType.MakeArrayType()
+					: elementType.MakeArrayType(rank);
+			}
+
+			if (parameter.IsByRef)
+			{
+				var elementType = GetClosedParameterType(parameter.GetElementType());
+				return elementType.MakeByRefType();
+			}
+
+			return parameter;
+
+			bool CloseGenericParametersIfAny(Type[] arguments)
+			{
+				var hasAnyGenericParameters = false;
+				for (var i = 0; i < arguments.Length; i++)
 				{
-					types.Add(genType);
+					var newType = GetClosedParameterType(arguments[i]);
+					if (newType != null && !ReferenceEquals(newType, arguments[i]))
+					{
+						arguments[i] = newType;
+						hasAnyGenericParameters = true;
+					}
 				}
+				return hasAnyGenericParameters;
 			}
+		}
+
+		public Type GetGenericArgument(int position)
+		{
+			Debug.Assert(0 <= position && position < genericTypeParams.Length);
 
-			return types.ToArray();
+			return genericTypeParams[position];
```

Comment:
I'm happy with just an assertion, since this is fixing the actual bug of using generic type parameter names when names are not part of the contract, we shouldn't get into this situation again anyway.

Happy to leave the question regarding `ClassEmitter` subclasses to you. Would definitely be great to see open generics, but not sure when that'll happen.